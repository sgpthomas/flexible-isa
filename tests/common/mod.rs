use std::{
    fmt::Display,
    path::{Path, PathBuf},
};

use flexible_isa::{cli, run, HalideExprOp, InstructionSelect};
use itertools::Itertools;

const BENCHMARKS_PATH: &str = "cache";
#[allow(unused)]
const BENCHMAKRS: &[&str] = &[
    "add",
    "average_pool",
    "blur3x3",
    "camera_pipe",
    "conv3x3a16",
    "conv3x3a32",
    "conv_nn",
    "depthwise_conv",
    "dilate3x3",
    // "elementwise",
    // "fully_connected",
    "gaussian3x3",
    "gaussian5x5",
    "gaussian7x7",
    "l2norm",
    "matmul",
    "max_pool",
    "mean",
    "median3x3",
    "mul",
    "sobel3x3",
    "softmax",
];

pub fn test_single_isa<S: AsRef<Path> + Display>(name: S) -> anyhow::Result<()> {
    let mut path = PathBuf::from(BENCHMARKS_PATH).join(&name);
    path.set_extension("stmt");

    let args: cli::Args = cli::Args::new(&[path]).learn(true);
    let isa = run(args)?;

    let instr_hist = InstructionSelect::from_recexpr(&isa.expressions);
    let hist: Vec<_> = instr_hist
        .iter()
        .filter_map(|(op, count)| {
            if let HalideExprOp::Instruction(i) = op {
                Some((i, count))
            } else {
                None
            }
        })
        // sort first by count, then by index
        .sorted_by_key(|(i, count)| (*count, *i))
        .map(|(i, count)| (i, count, isa.instructions[i].to_string()))
        .collect();

    insta::with_settings!({
        description => format!("generate isa for {name}"),
        omit_expression => true,
        snapshot_suffix => format!("{name}")
    }, {
        insta::assert_debug_snapshot!(hist)
    });

    Ok(())
}
