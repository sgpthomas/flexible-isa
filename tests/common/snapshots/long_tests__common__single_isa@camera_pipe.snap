---
source: tests/common/mod.rs
description: generate isa for camera_pipe
---
(
    BestIsa {
      2: (neg ?x0)
      306: (+ ?x0 ?x1)
      356: (- ?x0 ?x1)
      492: (* ?x0 ?x1)
      507: (/ ?x0 ?x1)
      519: (% ?x0 ?x1)
      524: (< ?x0 ?x1)
      526: (<= ?x0 (cast<unknown> ?x1))
      528: (== (cast<int32> ?#0) (cast<int32> ?#1))
      529: (&& (cast<bool> ?#0) (cast<bool> ?#1))
      530: (:: ?#0 ?#1)
      555: (get ?x0 ?x1)
    }
    ,
    [
        (
            i528,
            1,
            "(== (cast<int32> ?#0) (cast<int32> ?#1))",
        ),
        (
            i529,
            1,
            "(&& (cast<bool> ?#0) (cast<bool> ?#1))",
        ),
        (
            i530,
            2,
            "(:: ?#0 ?#1)",
        ),
        (
            i526,
            5,
            "(<= ?x0 (cast<unknown> ?x1))",
        ),
        (
            i524,
            7,
            "(< ?x0 ?x1)",
        ),
        (
            i2,
            46,
            "(neg ?x0)",
        ),
        (
            i356,
            80,
            "(- ?x0 ?x1)",
        ),
        (
            i519,
            99,
            "(\"%\" ?x0 ?x1)",
        ),
        (
            i555,
            191,
            "(get ?x0 ?x1)",
        ),
        (
            i507,
            385,
            "(/ ?x0 ?x1)",
        ),
        (
            i492,
            627,
            "(* ?x0 ?x1)",
        ),
        (
            i306,
            786,
            "(+ ?x0 ?x1)",
        ),
    ],
)
